homeassistant/core.py

```diff
@@ -218,7 +218,10 @@ def async_add_job(
         if asyncio.iscoroutine(target):
             task = self.loop.create_task(target)
         elif is_callback(target):
-            self.loop.call_soon(target, *args)
+            async def wrapper() -> Any:
+                """Wrapper coroutine around a callback."""
+                return target(*args)
+            task = self.loop.create_task(wrapper())
```

Comment:
Don't wrap it into a slow coroutine.

Reply:
Any suggestions how to properly do that?

1) If the caller knows that the target is a @callback and is not afraid of exceptions - they can call it directly.
2) If the caller knows that the target is a @callback, they are afraid of exceptions and don't care about return value - they can call `loop.call_soon` manually (i.e. the previous implementation)
3) If the caller knows that the target is a @callback, they are afraid of exceptions and they want the return value - the old implementation didn't provide that, the new one does.

4) If the caller doesn't know what `target` is and they need the return value - same as (3)

5) If the caller doesn't know what `target` is and they don't need the return value - I agree that the old implementation is better, but it would be also better for coro not to be wrapped in task for such a case.


Generally having a single function with different return types is bad.

Reply:
> Generally having a single function with different return types is bad.
For typing, yes but not for python itself

This change is very bad for the core and extend the memory footprint. We should hold the core slim and don't blow it up for a not exists type checking. Python is not strict and I hope that will be change in 4.x like c++ but actually I see no reason to make this bad change. In that case, I dislike this change.

Reply:
On `dev`

```
› hass --script benchmark async_million_state_changed_helper
Using event loop: asyncio.unix_events
Benchmark async_million_state_changed_helper done in 4.237771602987777s
Benchmark async_million_state_changed_helper done in 4.250693870999385s
Benchmark async_million_state_changed_helper done in 4.256614304991672s
Benchmark async_million_state_changed_helper done in 4.24649005298852s
```

With this PR:

```
› hass --script benchmark async_million_state_changed_helper
Using event loop: asyncio.unix_events
Benchmark async_million_state_changed_helper done in 5.726596449007047s
Benchmark async_million_state_changed_helper done in 5.9516269309970085s
Benchmark async_million_state_changed_helper done in 5.672646453007474s
Benchmark async_million_state_changed_helper done in 5.7938863370072795s
```

That's just a quick check of running it a couple of times. That's a significant slowdown.

Reply:
How do you relate the million events benchmark to real world use? The slowdown here seems to be 35% which sounds a lot. But it's also important how far from a noticeable impact we are in a real world use case. A million events in 5.8 seconds is 173k events per second. So if we fire more events than that, the clock will start to lag, correct?

Reply:
The problem with performance is that it's often a death by a thousand cuts. This gets a little slower, than that gets a little slower etc. Having a core piece of code become slower will make everything a little slower. 

I did a quick implementation where I replaced all `callback` functions used for the benchmark with async functions as Pascal suggested in final paragraph of his comment. That makes it even slower:

```
› hass --script benchmark async_million_state_changed_helper
Using event loop: asyncio.unix_events
Benchmark async_million_state_changed_helper done in 14.758733441005461s
Benchmark async_million_state_changed_helper done in 15.096011087996885s
Benchmark async_million_state_changed_helper done in 15.226054263010155s
```

Reply:
(this converted both the listener inside the benchmark as the function defined inside async_track_state_change to async functions)

Reply:
Interestingly, if I just convert the function inside track_stae_change to be async, I get back to ~5.5 seconds per run

Reply:
I think a proper solution is to make 5 new (type-safe) functions for all the usecases I mentioned above, then convert the callers and then delete the functions that are not actually used

Reply:
I agree with that approach.

Reply:
Hmm. That is python and not type safe. To make the code complex for nothing is not realy cool. If that is a c++ project, I would agree but not with python...